File: 1704. Determine if String Halves Are Alike/halves_are_alike.rs
use std::collections::{HashSet, HashMap};
use std::iter::FromIterator;

impl Solution {
    
    // O(n) time,
    // O(1) space,
    // Approach: string, array, counting
    pub fn halves_are_alike(s: String) -> bool {
        
        let v = [ 'A', 'E', 'I', 'O', 'U', 'a', 'e', 'i', 'o', 'u' ];
        
        let vowels: HashSet<&char> = HashSet::from_iter(v.iter());
        let mut vowel_sum = 0;
        
        let mid = s.len()/2;
        
        for (i, ch) in s.chars().enumerate() {            
            if !vowels.contains(&ch) {
                continue;
            }
            
            if i < mid {
                vowel_sum += 1;
            }
            else {
                vowel_sum -=1;
            }
        }
        
        return vowel_sum == 0;
    }
}

File: 1704. Determine if String Halves Are Alike/HalvesAreAlike.py
class Solution:
    
    # O(n) time,
    # O(1) space,
    # Approach: counting, 
    def halvesAreAlike(self, s: str) -> bool:
        vowels = { 'a', 'e', 'i', 'o', 'u' }
        count = 0
        
        for i, ch in enumerate(s):
            if ch.lower() in vowels:
                if i < (len(s)//2):
                    count += 1
                else:
                    count -= 1

        return count == 0